packages/muelu/test/unit_tests/NullspaceFactory.cpp

```diff
@@ -0,0 +1,149 @@
+
+#include "MueLu_config.hpp"
+
+#include "MueLu_TestHelpers.hpp"
+#include "MueLu_Version.hpp"
+
+#include <Xpetra_MultiVectorFactory.hpp>
+#include <MueLu_VerbosityLevel.hpp>
+#include "MueLu_Exceptions.hpp"
+#include "MueLu_Utilities.hpp"
+#include "MueLu_NullspaceFactory.hpp"
+//#include "MueLu_FactoryManager.hpp"
+#include "MueLu_CreateXpetraPreconditioner.hpp"
+
+namespace MueLuTests {
+
+
+  TEUCHOS_UNIT_TEST_TEMPLATE_4_DECL(NullspaceFactory,3DElasticity, Scalar, LocalOrdinal, GlobalOrdinal, Node)
+  {
+#   include "MueLu_UseShortNames.hpp"
+    MUELU_TESTING_SET_OSTREAM;
+    MUELU_TESTING_LIMIT_SCOPE(Scalar,GlobalOrdinal,Node);
+    out << "version: " << MueLu::Version() << std::endl;
+
+    using TST                   = Teuchos::ScalarTraits<SC>;
+    using magnitude_type        = typename TST::magnitudeType;
+    using TMT                   = Teuchos::ScalarTraits<magnitude_type>;
+    using real_type             = typename TST::coordinateType;
+    using RealValuedMultiVector = Xpetra::MultiVector<real_type,LO,GO,NO>;
+
+    Level fineLevel, coarseLevel;
+    TestHelpers::TestFactory<Scalar, LO, GO, NO>::createTwoLevelHierarchy(fineLevel, coarseLevel);
+    RCP<const Teuchos::Comm<int> > comm = Parameters::getDefaultComm();
+
+    GO nx = 2, ny = 3, nz = 4, mx = comm->getSize(), my = 1, mz = 1; ;
+    LO blkSize = 3;
+    Teuchos::ParameterList matrixList;
+    matrixList.set("nx", nx);
+    matrixList.set("ny", ny);
+    matrixList.set("nz", nz);
+    matrixList.set("mx", mx);
+    matrixList.set("my", my);
+    matrixList.set("mz", mz);
+    matrixList.set("matrixType","Elasticity3D");
+    RCP<Matrix> Op = TestHelpers::TestFactory<Scalar, LO, GO, NO>::BuildMatrix(matrixList,TestHelpers::Parameters::getLib());
+    Op->SetFixedBlockSize(blkSize);
+
+    fineLevel.Set("A",Op);
+
+    Teuchos::ParameterList galeriList;
+    galeriList.set("nx", nx);
+    galeriList.set("ny", ny);
+    galeriList.set("nz", nz);
+    RCP<Matrix> fake = TestHelpers::TestFactory<Scalar, LO, GO, NO>::Build1DPoisson(nx*ny*nz); // need scalar PDE to produce proper coordinate array
+    RCP<RealValuedMultiVector> coordinates = Galeri::Xpetra::Utils::CreateCartesianCoordinates<SC,LO,GO,Map,RealValuedMultiVector>("3D", fake->getRowMap(), galeriList);
+    fineLevel.Set("Coordinates", coordinates);
+
+    RCP<NullspaceFactory> nspace = rcp(new NullspaceFactory());
+    fineLevel.Request("Nullspace", &(*nspace));
+    Teuchos::ParameterList nspaceList = *(nspace->GetValidParameterList());
+    nspaceList.set("nullspace: calculate rotations", true);
+    nspace->SetParameterList(nspaceList);
+    fineLevel.Request(*nspace);
+    nspace->Build(fineLevel);
+    RCP<MultiVector> nullSpace = fineLevel.Get<RCP<MultiVector> >("Nullspace",nspace.get());
+
+    ArrayRCP<real_type> xvals = coordinates->getDataNonConst(0); 
+    ArrayRCP<real_type> yvals = coordinates->getDataNonConst(1); 
+    ArrayRCP<real_type> zvals = coordinates->getDataNonConst(2); 
```

Comment:
This could be const:
```suggestion
    ArrayRCP<real_type> xvals = coordinates->getData(0); 
    ArrayRCP<real_type> yvals = coordinates->getData(1); 
    ArrayRCP<real_type> zvals = coordinates->getData(2); 
```